charmhelpers/core/hookenv.py

```diff
@@ -1135,6 +1135,28 @@ def network_get(endpoint, relation_id=None):
     return yaml.safe_load(response)
 
 
+def ingress_address(rid=None, unit=None):
+    """
+    Retrieve the ingress-address from a relation when available. Otherwise,
+    return the private-address. This function is to be used on the consuming
+    side of the relation.
+
+    Usage:
+    addresses = []
+    for rid in relation_ids(relation_name):
+        for unit in related_units(rid):
+             addresses.append(ingress_address(rid=rid, unit=unit))
```

Comment:
Not a problem here, but something that really does annoy me, is that number of times I see this pattern in code; it's crying out for a helper.

e.g.

```python

RelationAddr = NamedTuple('RelationAddr', 'rid unit')

def iter_units_for_relation_name(relation_name):
    for rid in relation_ids(relation_name):
        for unit in related_units(rid):
            yield RelationAddr(rid, unit)

# usage
addresses = [ingress_address(rid=u.rid, unit=u.unit)
             for u in iter_units_for_relation_name(name)]
```

I feel like I need to add this to charmhelpers ...

Reply:
I totally agree. I actually think juju should provide a simple:
relation-get --all -r RID
But that is wishful thinking.
I am +1 on your generator.